clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -411,9 +415,36 @@ public void closeVeniceWriters(boolean doFlush) {
   }
 
   @Override
-  protected void closeVeniceViewWriters() {
+  protected void closeVeniceViewWriters(boolean doFlush) {
     if (!viewWriters.isEmpty()) {
-      viewWriters.forEach((k, v) -> v.close());
+      long gracefulCloseDeadline = time.getMilliseconds() + VIEW_WRITER_CLOSE_TIMEOUT_IN_MS;
+      viewWriters.forEach((k, v) -> v.close(doFlush));
+      // Short circuit last VT produce call future if it's incomplete to unblock any consumer thread(s) that are waiting
+      for (PartitionConsumptionState pcs: partitionConsumptionStateMap.values()) {
+        CompletableFuture<Void> lastVTProduceCallFuture = pcs.getLastVTProduceCallFuture();
+        if (!lastVTProduceCallFuture.isDone()) {
+          if (doFlush) {
+            long timeout = gracefulCloseDeadline - time.getMilliseconds();
```

Comment:
so the timeout here is aggragate pcs map closing timeout? or we want to measure against each of the last produce future timeout ?

Reply:
my point was that say we have 10 pending lastVTProduceCallFuture, and first two took 59s and then you give only 1s for the rest of the futures. ie the concern is timeout in the following code is extremely non-uniform `lastVTProduceCallFuture.get(timeout, MILLISECONDS);`

Reply:
I just realized that waiting `lastVTProduceCallFuture` to finish could really be a heavy wait, it means to wait all its previous messages and view messages to get done.

Also, does it make sense to declare lastVTProduceCallFuture as `volatile`? It can be set and get in different threads and we don't really want to complete a stale future?
`private CompletableFuture<Void> lastVTProduceCallFuture;`

Reply:
This is why we short-circuit the entire chain after waiting it for some timeout based on the overall deadline given to close all view writers.
Regarding to your example, if first 2 took 59s it's not like the remaining 8 lastVTProduceCallFuture depends on the first 2. They also had 59s to finish and if they are still not done yet and the deadline is 60s I'm only going to wait for one more second before I short-circuit and complete them exceptionally.